### DISCUSSION TITLE
Load Balance Connection

### DISCUSSION THREAD

**Discussion by Purnajith** [rabbitmq/rabbitmq-stream-dotnet-client#188] [Q&A] [answered]
We have a 3-node rabbitMQ cluster ( Deployed on EC2 machines with names as rmq1.internal, rmq2.internal, rmq3.internal ) listening on target groups via the network load balancer. We are planning to provide the load balancer's DNS record which would resolve to the RMQ cluster. But when we tried using it always provided an error as  "RabbitMQ.Stream.Client.RoutingClientException: Could not find broker (rmq2.internal:5551) after 25 attempts" 

Here is the sample code 
```
var ipAddress = Dns.GetHostAddresses("rmq.internal");
var lbAddressResolver = new AddressResolver(new IPEndPoint(IPAddress.Parse(ipAddress), 5551));
var config = new StreamSystemConfig
{
    UserName = "guest",
    Password = "guest",
    VirtualHost = "/",
    AddressResolver = lbAddressResolver,
    Endpoints = new List<EndPoint> {addressResolver.EndPoint},
}; 
```

Can you help us in resolving this issue ?

FYI 
-  We see this issue only in the dot net stream lib, not in either dotnet queue lib [lib](https://www.rabbitmq.com/dotnet.html) or the java libs (both queue and stream) which is using the same hostname and directly connecting to the load balancer address resolver  as defined [here ](https://blog.rabbitmq.com/posts/2021/07/connecting-to-streams/#using-the-stream-java-client-with-a-load-balancer)

**Comment by Gsantomaggio** [comment]
Port `5551` is a TLS port. You need to configure the TLS client side:
https://github.com/rabbitmq/rabbitmq-stream-dotnet-client#tls

**Comment by Purnajith** [comment]
We have enabled that as well with the following 

```
Ssl = new RSC.SslOption
                {
                    Enabled = useSSL,
                    ServerName = hostName,
                },
```

**Reply by Gsantomaggio** [reply]
I don't think that it is enough as a configuration. 

btw, can you please check the rabbitmq server logs?  you should see some connection.

Another test is to try without TLS using the standard port `5552`

**Reply by Purnajith** [reply]
Ohh will try that and let you know

**Comment by Gsantomaggio** [comment]
I just tried with a configuration like this:
```csharp
  var ssl = new SslOption()
        {
            Enabled = true,
            AcceptablePolicyErrors = SslPolicyErrors.RemoteCertificateNotAvailable |
                                     SslPolicyErrors.RemoteCertificateChainErrors |
                                     SslPolicyErrors.RemoteCertificateNameMismatch,
        };
        var ipAddress = await Dns.GetHostAddressesAsync("my.domain.net");
        var lbAddressResolver = new AddressResolver(new IPEndPoint(ipAddress[0], 5551));

        var config = new StreamSystemConfig()
        {
            UserName = "myusername",
            Password = "mypassword",
            Ssl = ssl,
            AddressResolver = lbAddressResolver,
            Endpoints = new List<EndPoint>(new List<EndPoint>()
            {
                lbAddressResolver.EndPoint
            }),
        };
```

and it works. 
But I think that I should add a DNS entry for the `AddressResolver` to avoid to lookup manually the IP Address from the DNS

**Reply by Purnajith** [reply]
@Gsantomaggio Great this is help full, I will try it and let you know

**Reply by Gsantomaggio** [reply]
@Purnajith, FYI  https://github.com/rabbitmq/rabbitmq-stream-dotnet-client/pull/189

In this way, you can use the DNS directly

**Comment by Gsantomaggio** [comment]
@Purnajith have you solved it?  We will release a new version soon I'd like to know if everything is ok!
thank you

**Reply by Purnajith** [reply]
@Gsantomaggio We were not able to check cause we rolled back out the set up for a single cluster so that no one would be blocked.  I'll try to test this by this Friday and let you know if is that ok? 

When are you planning to release a new version?

**Reply by Gsantomaggio** [reply]
Thank you @Purnajith ! the plan is for the next week

**Reply by Purnajith** [reply]
Great will  try to give you an update before that

**Comment by Gsantomaggio** [comment]
will close due of inactivity

**Comment by lukas8219** [comment]
Hi @Gsantomaggio , When using CloudAMQP-hosted instances with 3 nodes we experience the same `RabbitMQ.Stream.Client.RoutingClientException: Could not find broker`.  On CloudAMQP the load balancing is done via DNS A records. I tried all the options here and also didn't work.

I solved it eventually by hardcoding `DnsEndpoint` such as below:

```dotnet
var _config = new StreamSystemConfig
        {
            Endpoints = [
                new DnsEndPoint("test-crisp-dark-rhino-01.rmq7.cloudamqp.com", 5552),
                new DnsEndPoint("test-crisp-dark-rhino-02.rmq7.cloudamqp.com", 5552),
                new DnsEndPoint("test-crisp-dark-rhino-03.rmq7.cloudamqp.com", 5552)
            ],
            UserName = defaultUser,
            Password = <pass>,
            VirtualHost = defaultUser,
            ClientProvidedName = "test",
        };
```

My investigation showed up that the `DnsEndpoint + AddressResolver` would use it as hostname, mismatching the advertised_host etc.

I wonder if this the expected way or we could offer another way around?

You can repro the issue by:
- Spin 3 node instance on CloudAMQP
- Try to connect to the cluster stream with the below snippet

```
var streamName = "test_stream";
var endpoint = new DnsEndPoint("test-crisp-dark-rhino.rmq7.cloudamqp.com", 5552);
var addressResolver = new AddressResolver(endpoint);

var _config = new StreamSystemConfig
        {
            AddressResolver = addressResolver,
            Endpoints = [endpoint],
            UserName = "test",
            Password = "***",
            VirtualHost = "test",
            ClientProvidedName = "test",
        };
```

Regards,

**Reply by Gsantomaggio** [reply]
Hi @lukas8219,
I'm not familiar with cloudAMQP. Are you sure the RabbitMQ stream plugin is enabled, and that port 5552 is exposed? 
5552 is the standard plain port and it is strange.
5551 is the TLS port. 

Please check if the TLS is open and the [Stream Plugin is enabled](https://www.rabbitmq.com/docs/stream)

**Reply by lukas8219** [reply]
Hi @Gsantomaggio Yes, plugins work as expected. The issue only happen when theres more than 1 node ie being load balanced

**Reply by Gsantomaggio** [reply]
@lukas8219 Can you please try https://github.com/rabbitmq/rabbitmq-stream-dotnet-client/pull/449 I added some logs. and report what the [logs say](https://github.com/rabbitmq/rabbitmq-stream-dotnet-client/pull/449/changes#diff-8f78f86ce0aae80f0a305a68a38cf18269bf81fa53428b84351dcea72a0a3077R98-R100)

**Reply by Gsantomaggio** [reply]
You should see something like:
```
[14:59:17] dbug: RabbitMQ.Stream.Client.Reliable.Consumer[0] advertised_host or advertised_port doesn't match. Expected: rabbitmq-server-0.rabbitmq-nodes.default:5552, Actual: rabbitmq-server-1.rabbitmq-nodes.default:5552. Attempt number: 1/25
```

Questions:

- Don't you see any connection in the magament ui? 
- Do you see something in the RabbitMQ logs?

**Reply by michaelklishin** [reply, marked as answer]
@Gsantomaggio honestly, I suggest that you let @lukas8219 or someone else from CloudAMQP to debug this behavior and contribute. They do not contribute virtually anything to RabbitMQ but directly benefit financially from every RabbitMQ and client library improvement we make.

**Comment by michaelklishin** [comment]
@lukas8219 wy should the core team spend time investigating what is preventing [stream protocol connections](https://www.rabbitmq.com/docs/stream-connections) in a CloudAMQP-hosted cluster?

If you work for CloudAMQP or pay them, don't expect the core team to investigate your issues. CloudAMQP directly make money off of RabbitMQ all while contributing less than 1% of R&D.

Gabriele is being extremely generous here.
